swarm/src/protocols_handler/multi.rs

```diff
@@ -0,0 +1,260 @@
+    where
+        T: IntoIterator<Item = (K, H)>
+    {
+        Handler { handlers: HashMap::from_iter(iter) }
+    }
+}
+
+impl<K, H> ProtocolsHandler for Handler<K, H>
+where
+    K: Clone + std::fmt::Debug + Hash + Eq + Send + 'static,
+    H: ProtocolsHandler,
+    H::InboundProtocol: InboundUpgradeSend,
+    H::OutboundProtocol: OutboundUpgradeSend
+{
+    type InEvent = (K, <H as ProtocolsHandler>::InEvent);
+    type OutEvent = (K, <H as ProtocolsHandler>::OutEvent);
+    type Error = <H as ProtocolsHandler>::Error;
+    type InboundProtocol = Upgrade<K, <H as ProtocolsHandler>::InboundProtocol>;
+    type OutboundProtocol = <H as ProtocolsHandler>::OutboundProtocol;
+    type OutboundOpenInfo = (K, <H as ProtocolsHandler>::OutboundOpenInfo);
+
+    fn listen_protocol(&self) -> SubstreamProtocol<Self::InboundProtocol> {
+        let upgrades = self.handlers.iter()
+            .map(|(k, h)| (k.clone(), h.listen_protocol().into_upgrade().1))
+            .collect();
+        SubstreamProtocol::new(Upgrade { upgrades })
+    }
+
+    fn inject_fully_negotiated_outbound (
+        &mut self,
+        protocol: <Self::OutboundProtocol as OutboundUpgradeSend>::Output,
+        (key, arg): Self::OutboundOpenInfo
+    ) {
+        if let Some(h) = self.handlers.get_mut(&key) {
+            h.inject_fully_negotiated_outbound(protocol, arg)
+        } else {
+            log::error!("inject_fully_negotiated_outbound: no handler for key {:?}", key)
+        }
+    }
+
+    fn inject_fully_negotiated_inbound (
+        &mut self,
+        (key, arg): <Self::InboundProtocol as InboundUpgradeSend>::Output
+    ) {
+        if let Some(h) = self.handlers.get_mut(&key) {
+            h.inject_fully_negotiated_inbound(arg)
+        } else {
+            log::error!("inject_fully_negotiated_inbound: no handler for key {:?}", key)
+        }
+    }
+
+    fn inject_event(&mut self, (key, event): Self::InEvent) {
+        if let Some(h) = self.handlers.get_mut(&key) {
+            h.inject_event(event)
+        } else {
+            log::error!("inject_event: no handler for key {:?}", key)
+        }
+    }
+
+    fn inject_dial_upgrade_error (
+        &mut self,
+        (key, arg): Self::OutboundOpenInfo,
+        error: ProtocolsHandlerUpgrErr<<Self::OutboundProtocol as OutboundUpgradeSend>::Error>
+    ) {
+        if let Some(h) = self.handlers.get_mut(&key) {
+            h.inject_dial_upgrade_error(arg, error)
+        } else {
+            log::error!("inject_dial_upgrade_error: no handler for protocol {:?}", key)
+        }
+    }
+
+    fn connection_keep_alive(&self) -> KeepAlive {
+        self.handlers.values()
+            .map(|h| h.connection_keep_alive())
+            .max()
+            .unwrap_or(KeepAlive::No)
+    }
+
+    fn poll(&mut self, cx: &mut Context)
+        -> Poll<ProtocolsHandlerEvent<Self::OutboundProtocol, Self::OutboundOpenInfo, Self::OutEvent, Self::Error>>
+    {
+        for (k, h) in self.handlers.iter_mut() {
```

Comment:
This might starve the last handler in case any handler other than the last returns `Poll::Ready` for each call to `poll`, right? Is that an issue or negligible given its low probability?

Reply:
True, with larger numbers of handlers this might become a problem.

Reply:
NB that no matter in which order we perform the polling starvation always becomes a problem with a larger number of handlers.